### ISSUE TITLE
lack of reproducibility despite using the same seed

### ISSUE THREAD

**Issue by deanforce** [AnotherSamWilson/ParBayesianOptimization#43]
Hello,
When trying to use BayesianOptimization to optimize choices for hyperparameters for an xgboost model, I noticed that despite using an identical seed, I get different results on repeated runs. I am providing below some sample code with toy data that reproduces the problem. When reviewing the output dataframes from the two identical runs (‘a’ and ‘b’), one can notice that these runs lead to consistent results for the initial parameters chosen during the design phase, but that once the optimization phase begins the results from the two runs diverge. In the actual data, this often results in different optimal hyperparameters being chosen. Why would this be happening despite using the same seed? Thank you!

```
db<-read.delim("https://drive.google.com/uc?export=download&id=1JXMUbwAWi8jjRPGQxzC15tUzSIWrPfNH")
db<-as.data.frame(db)

library("xgboost")
library("data.table")
library("skimr")
library("DiceKriging")
library("ParBayesianOptimization")
library("Matrix")


sparse_matrix<- sparse.model.matrix(outcome ~ .-1, data = db)
labels <- db$outcome
db_xgb <- xgb.DMatrix(data = sparse_matrix, label= labels)


scoringFunction <- function(max_depth, min_child_weight, eta,gamma,max_delta_step,alpha,lambda,subsample) {
    set.seed(1234, "L'Ecuyer-CMRG")
    xgbcv <- xgb.cv(params = list(
      nthreads=NULL,
      max_depth = max_depth,
      min_child_weight = min_child_weight,
      eta=eta,
      gamma=gamma,
      max_delta_step=max_delta_step,
      alpha=alpha,
      lambda=lambda,
      subsample = subsample,
      booster="gbtree",
      objective = "binary:logistic",  # for regression models
      eval_metric="logloss"),
      tree_method="hist",
      nround=10000,
      nfold=5,
      early_stopping_rounds = 50,
      verbose=0, 
      maximize=F,
      data = db_xgb)
    
    xgbcv$evaluation_log[,min(xgbcv$evaluation_log$test_logloss_mean)]
    
    return(
      list( 
        Score = -min(xgbcv$evaluation_log$test_logloss_mean)
        , nrounds = xgbcv$best_iteration
      )
    )
  }

  bounds <- list( 
    max_depth = c(2L, 40L),
    min_child_weight=c(1L,30L),
    gamma=c(0,10),
    eta=c(0.0001,1),
    max_delta_step=c(0,20),
    alpha=c(0,7),
    lambda=c(0,7),
    subsample=c(0.5,1))
  
#first run  
  set.seed(1234)
  optObj_a <- bayesOpt(
    FUN = scoringFunction
    , bounds = bounds
    , initPoints = 40
    , iters.n = 20
  )
  
a<-optObj_a$scoreSummary

#second run, with identical seed
set.seed(1234)
optObj_b <- bayesOpt(
  FUN = scoringFunction
  , bounds = bounds
  , initPoints = 40
  , iters.n = 20
)

b<-optObj_b$scoreSummary

```

**Comment by AnotherSamWilson** [comment]
If you set the seed in the xgboost parameters does this fix the issue?

**Comment by deanforce** [comment]
Thanks for the reply. So passing the seed as part of the xgboost parameters doesn't seem to make a difference in this regard - one still observes a discrepancy between the repeated runs, which sometimes leads to meaningful differences. Any thoughts on what might be causing this behavior?

**Comment by AnotherSamWilson** [comment]
Weird, it looks like runs 7, 9 and 15 didn't return the same scores in each run. They are barely different, but still different. This causes a different GP to be trained. You can see this if you run `a == b` in your code. I don't know why xgboost would be have this way, I don't really use that package.
